juju/juju
# [JUJU-4353] Accept base and series in bundles if they match

Add a verification step into the client to ensure that series and base
match.

As such, we can treat series and bases as equivalent when they show up
precedence-wise.

Immediately combine into a base in core. As a helpful side effect, this
means we can minimise our dealing with series in this part of the code,
which will make it very easy to remove completely

## Checklist


## QA steps

Create 3 bundles as follows
```
$ cat bundle1.yaml
applications:
  ubu1:
    charm: ubuntu
    base: ubuntu@22.04
    series: focal
    num_units: 1

$ cat bundle2.yaml
default-base: ubuntu@22.04
applications:
  ubu2:
    charm: ubuntu
    base: ubuntu@20.04
    series: focal
    num_units: 1

$ cat bundle3.yaml
default-base: ubuntu@22.04
applications:
  ubu3:
    charm: ubuntu
    series: focal
    num_units: 1
```

Verify:
```
$ juju deploy ./bundle1.yaml
ERROR cannot deploy bundle: supplied series "focal" and base "ubuntu@22.04" in application "ubu1" do not match not valid
```
```
$ juju deploy ./bundle2.yaml
Located charm "ubuntu" in charm-hub, channel stable
Executing changes:
- upload charm ubuntu from charm-hub for base ubuntu@20.04/stable with architecture=amd64
- deploy application ubu2 from charm-hub on ubuntu@20.04/stable using ubuntu
- add unit ubu2/0 to new machine 0
Deploy of bundle completed.


$ juju status
Model  Controller  Cloud/Region         Version      SLA          Timestamp
m1     lxd         localhost/localhost  3.3-beta2.1  unsupported  11:41:06+01:00

App   Version  Status   Scale  Charm   Channel  Rev  Exposed  Message
ubu2           waiting    0/1  ubuntu  stable    24  no       waiting for machine

Unit    Workload  Agent       Machine  Public address  Ports  Message
ubu2/0  waiting   allocating  0        10.219.211.81          waiting for machine

Machine  State    Address        Inst id        Base          AZ  Message
0        pending  10.219.211.81  juju-41adf7-0  ubuntu@20.04      Running
```
```
$ juju deploy ./bundle3.yaml
Located charm "ubuntu" in charm-hub, channel stable
Executing changes:
- upload charm ubuntu from charm-hub for base ubuntu@20.04/stable with architecture=amd64
- deploy application ubu3 from charm-hub on ubuntu@20.04/stable using ubuntu
- add unit ubu3/0 to new machine 0
Deploy of bundle completed.

$ juju status
Model  Controller  Cloud/Region         Version      SLA          Timestamp
m2     lxd         localhost/localhost  3.3-beta2.1  unsupported  11:41:29+01:00

App   Version  Status   Scale  Charm   Channel  Rev  Exposed  Message
ubu3           waiting    0/1  ubuntu  stable    24  no       waiting for machine

Unit    Workload  Agent       Machine  Public address  Ports  Message
ubu3/0  waiting   allocating  0        10.219.211.163         waiting for machine

Machine  State    Address         Inst id        Base          AZ  Message
0        pending  10.219.211.163  juju-ff74a2-0  ubuntu@20.04      Running

```
